### PR TITLE
[Discussion] First pass for turning permissions into WP roles / caps

### PR DESCRIPTION
**Pull request by javorszky**
Ref: #19 

This is rebased on top of latest `wordpress-plugin-dbi:master` (https://github.com/deliciousbrains/GlotPress/commit/9abf742a09c62a941b09d0eb524fdc25c3b1433a) and contains what @A5hleyRich did.

This is just to look at code on Tuesday, 3rd Nov 2015, and have something to talk about. I'll share some of the thinking behind the code.
## WP Roles

In a conversation on the issue and in Slack (starting at https://wordpress.slack.com/archives/glotpress/p1445911018000469), the idea is to use roles and caps to augment what GP is doing on itself.

On the surface it looks easy enough, for any action, we check internal permissions AND a wp role, and then decide. For example, if someone were to try and translate a locale, we'll check whether they're logged in, and then we'd check whether the user has the `translator` role assigned to them.

I've introduced the two roles, and `translator` is being applied to anyone who gets a new WP user.

The main problem I found with that is that a lot of the functionality in current GP is generic and is based on `GP_Thing`'s `find_one`, and adding in a WP role check won't help much.

For example `GP_Route::cannot_and_redirect` will check `GP_Route::can` which checks `GP_User::can`. There's no way to know whether a request is for a translator or a validator in there, as it's used for every user. So if we're checking for a wp role where the `cannot_and_redirect`s are called, we miss out on the redirect part if only the role has been revoked. In that case `$can` would still return true, but user doesn't have the role, so unless we go in circles and introduce yet another filter, we won't get the `_and_redirect` functionality.

Same with the `_and_forbidden`.

To work around that, we could have a list of actions: for example if `edit` or `approve` are coming through, we're checking for `validator` role as well, if `edit`, we're checking for `translator`. This isn't the best way to do this in my opinion, as we'd be adding more complexity to the system.
## Advanced permissions

Those are all generic filters that all hook into `GP_User::can`, so again, not a place to add validation.

This is a large issue. Most of the PR is code I chose not to implement because it'd make a mess of the existing system.
## tl;dr
- Using WP caps for the granular permissions GP has is not great as we can't efficiently query by a cap it being in the `usermeta` table in a serialised format.
- Using WP roles in addition to the current permission system is awkward as those can be added in a number of different places. The earliest being `gp_before_request`, and the latest being `GP_User::can`, and in between, depending which route, it's a variety of `cannot_and_*`, then `GP_Route::can` and then the user's can.

### CODE REVIEW

**Comment by toolstack** [comment]
I took a closer look at this over the weekend and came to the conclusion that Roles make no sense for GlotPress.  Their one benefit (being able to query a list of users with them) is not actually a benefit as it does the same "query the serialized data" tactic as we would do for Capabilities anyway.

Once that was dealt with, the way forward was to see how permissions were actually used in GP.  They're really only used for Projects, Glossaries and Transition Sets.  Transition sets being the only place that actually has a UI for them.

Anyway, after a bit of whacking at it, I've got something that mostly works over in https://github.com/deliciousbrains/GlotPress/tree/19-toolstack-poc-capabilities

It's biggest deficiency at moment is that I haven't updated the permissions page to use it to get the user list.

From a day to day functioning perspective, Capabilities don't make a significant impact on GlotPress and so it's a toss up as to which is better.  The one area that implementing Capabilities really get messy is in the permissions page, but that can be handled.

At this point, I see no benefit to using Roles/Capabilities and only downside, if WordPress handled them better there would be no doubt that we should use them.  However with the state they're in at moment and the fact we have a solution in place that works better for our needs, I see no compelling reason to try and switch to them.

**Comment by javorszky** [comment]
> At this point, I see no benefit to using Roles/Capabilities and only downside, if WordPress handled them better there would be no doubt that we should use them. However with the state they're in at moment and the fact we have a solution in place that works better for our needs, I see no compelling reason to try and switch to them.

Totally agree with that. I think having a better UI for permissions would help with the underlying need: to be able to easily assign / revoke permissions to users.
